### PR TITLE
[TEST]: 5208: Server42 RTT: HaFlow

### PR DESCRIPTION
**Pull request by yuliiamir**

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42HaFlowRttSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+        and: "The server42 stats for both sub-Flows in forward direction are not increased"
+        !flowStats.of(haFlow.subFlows.first().flowId).get(FLOW_RTT, FORWARD, Origin.SERVER_42).hasNonZeroValuesAfter(timeWhenMissingRuleIsDetected)
+        !flowStats.of(haFlow.subFlows.last().flowId).get(FLOW_RTT, FORWARD, Origin.SERVER_42).hasNonZeroValuesAfter(timeWhenMissingRuleIsDetected)
+
+        and: "The server42 stats for both sub-Flows in reverse direction are increased"
+        Wrappers.wait(STATS_FROM_SERVER42_LOGGING_TIMEOUT + WAIT_OFFSET) {
+            assert flowStats.of(haFlow.subFlows.first().flowId).get(FLOW_RTT, REVERSE, Origin.SERVER_42).hasNonZeroValuesAfter(timeWhenMissingRuleIsDetected)
+            assert flowStats.of(haFlow.subFlows.last().flowId).get(FLOW_RTT, REVERSE, Origin.SERVER_42).hasNonZeroValuesAfter(timeWhenMissingRuleIsDetected)
+        }
```
Comment (pkazlenka) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42HaFlowRttSpec.groovy:149]: Amount of TSDB calls can be reduced twice here as well.

**Review by pkazlenka — CHANGES_REQUESTED** [review]
Good, but some minor problems should be solved.

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy [hunk 2]
```diff
@@ -726,4 +727,52 @@ class SwitchHelper {
...
+            def s42Rules = northbound.get().getSwitchRules(sw.dpId).flowEntries.findAll {
+                it.cookie in  [SERVER_42_FLOW_RTT_OUTPUT_VLAN_COOKIE,
+                               SERVER_42_FLOW_RTT_OUTPUT_VXLAN_COOKIE]
+            }
+            assert requiredState ? (s42Rules.size() == amountOfS42Rules) : s42Rules.empty
+        }
+        return originalProps.server42FlowRtt
+    }
+
+    static void verifyAbsenceOfServer42FlowRttRules(def initialSwitchRtt) {
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy]: Done

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy [hunk 3]
```diff
@@ -726,4 +727,52 @@ class SwitchHelper {
...
+            })
+        }
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            def amountOfS42Rules = (sw.features.contains(NOVIFLOW_PUSH_POP_VXLAN)
+                    || sw.features.contains(KILDA_OVS_PUSH_POP_MATCH_VXLAN)) ? 2 : 1
+            def s42Rules = northbound.get().getSwitchRules(sw.dpId).flowEntries.findAll {
+                it.cookie in  [SERVER_42_FLOW_RTT_OUTPUT_VLAN_COOKIE,
+                               SERVER_42_FLOW_RTT_OUTPUT_VXLAN_COOKIE]
+            }
+            assert requiredState ? (s42Rules.size() == amountOfS42Rules) : s42Rules.empty
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy]: Updated

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy [hunk 4]
```diff
@@ -726,4 +727,52 @@ class SwitchHelper {
     static SwitchPropertiesDto getCachedSwProps(SwitchId switchId) {
         return getCachedAllSwProps().find { it.switchId == switchId }
     }
+
+    static def changeFlowRttSwitch(Switch sw, boolean requiredState) {
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy]: renamed to setServer42FlowRttForSwitch

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy [hunk 5]
```diff
@@ -726,4 +727,52 @@ class SwitchHelper {
...
+                initialSwitchRtt.keySet().eachParallel { sw ->
+                    assert northbound.get().getSwitchRules(sw.dpId).flowEntries.findAll {
+                        new Cookie(it.cookie).getType() in [CookieType.SERVER_42_FLOW_RTT_INPUT,
+                                                            CookieType.SERVER_42_FLOW_RTT_INGRESS]
+                    }.empty
+                }
+            }
+        }
+    }
+    static def revertToOriginSwitchSetup(def initialSwitchRtt) {
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy:789]: +

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42FlowRttSpec.groovy [hunk 6]
```diff
@@ -80,12 +75,14 @@ class Server42FlowRttSpec extends HealthCheckSpecification {
         def switchPair = switchPairFilter(switchPairs.all().withBothSwitchesConnectedToServer42()).random()
 
         when: "Set server42FlowRtt toggle to true"
-        def flowRttFeatureStartState = changeFlowRttToggle(true)
+        def flowRttToggleInitialState = northbound.featureToggles.server42FlowRtt
+        !flowRttToggleInitialState && northbound.toggleFeature(FeatureTogglesDto.builder().server42FlowRtt(true).build())
+        //not going to check rules on every switch in the system. sleep does the trick fine
+        sleep(3000)
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42FlowRttSpec.groovy]: the appropriate method has been added (check all switches with server42 support)

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42HaFlowRttSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        when: "Delete ingress server42 rule(s) related to the flow on the shared switch"
+        //if ha-flow doesn't have shared path, shared endpoint is y-point and has  server42 flow rtt ingress rule per sub-Flows
+        def cookiesToDelete = northbound.getSwitchRules(swT.shared.dpId).flowEntries.findAll {
+            new Cookie(it.cookie).getType() == CookieType.SERVER_42_FLOW_RTT_INGRESS
+        }.cookie
+        cookiesToDelete.each { northbound.deleteSwitchRules(swT.shared.dpId, it) }
+
+        then: "System detects missing rule on the shared switch"
+        Wrappers.wait(RULES_DELETION_TIME) {
+            assert northbound.validateSwitch(swT.shared.dpId).rules.missing.sort() == cookiesToDelete.sort()
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42HaFlowRttSpec.groovy:131]: will update after the new logic for switch validation/sync

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy [hunk 8]
```diff
@@ -726,4 +727,74 @@ class SwitchHelper {
...
+            if (isServer42FlowRttEnabled) {
+                assert actualRuleCookies.size() == expectedNumberOfS42Rules
+            } else {
+                assert actualRuleCookies.isEmpty()
+            }
+        }
+        return originalProps.server42FlowRtt
+    }
+
+    static List<FlowEntry> s42SwitchRules(SwitchId swId) {
```
Comment (pkazlenka) [inline src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/SwitchHelper.groovy:756]: What if we have methods `getServer42SwitchRules()` and `getExpectedServer42SwitchRules()` (this one would calculate expected rules based on switch properties) and compare them in tests?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42FlowRttSpec.groovy [hunk 9]
```diff
@@ -96,7 +97,10 @@ class Server42FlowRttSpec extends HealthCheckSpecification {
         }
 
         cleanup: "Revert system to original state"
-        revertToOrigin([flow], flowRttFeatureStartState, initialSwitchRtt)
+        flow && flowHelperV2.deleteFlow(flow.flowId) && switchHelper.verifyAbsenceOfServer42FlowRttRules(initialSwitchProps.keySet())
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/server42/Server42FlowRttSpec.groovy:100]: Due to the specificity of the test(verifyAbsenceOfServer42FlowRttRules), flow deletion should be executed as a first step of the cleanup section.

**Review by pablomuri — APPROVED** [review]

**Pull request merged**